File: skills/make-docx/scripts/smoke_create.mjs
#!/usr/bin/env node
/**
 * Smoke: create a sample .docx (with a DXA table) and run check_docx_tables.py.
 * Usage (from repo / workspace root):
 *   node skills/make-docx/scripts/smoke_create.mjs
 */
import { createRequire } from "node:module";
import { existsSync, mkdirSync, writeFileSync } from "node:fs";
import { dirname, join, resolve } from "node:path";
import { fileURLToPath } from "node:url";
import { spawnSync } from "node:child_process";

const __dirname = dirname(fileURLToPath(import.meta.url));
const skillRoot = resolve(__dirname, "..");
const workspaceHint = resolve(skillRoot, "../..");
const smokeNm = join(workspaceHint, "outputs", "_office-smoke", "node_modules", "docx", "package.json");
const require = createRequire(existsSync(smokeNm) ? smokeNm : import.meta.url);
const {
  Document,
  Packer,
  Paragraph,
  TextRun,
  HeadingLevel,
  Table,
  TableRow,
  TableCell,
  WidthType,
  BorderStyle,
  Header,
  Footer,
  PageNumber,
  AlignmentType,
} = require("docx");

const outDir = join(workspaceHint, "outputs", "documents", "docx");
const outFile = join(outDir, "smoke-q3-summary.docx");

const TABLE_W = 9026; // A4 content width @ 1" margins
const COLS = [3010, 3010, 3006];

function cell(text, width, opts = {}) {
  return new TableCell({
    width: { size: width, type: WidthType.DXA },
    borders: {
      top: { style: BorderStyle.SINGLE, size: 4, color: "B0B0B0" },
      bottom: { style: BorderStyle.SINGLE, size: 4, color: "B0B0B0" },
      left: { style: BorderStyle.SINGLE, size: 4, color: "B0B0B0" },
      right: { style: BorderStyle.SINGLE, size: 4, color: "B0B0B0" },
    },
    shading: opts.fill ? { fill: opts.fill } : undefined,
    children: [
      new Paragraph({
        children: [new TextRun({ text, bold: !!opts.bold, size: 20 })],
      }),
    ],
  });
}

const doc = new Document({
  styles: {
    default: { document: { run: { font: "Calibri", size: 24 } } },
    paragraphStyles: [
      {
        id: "Heading1",
        name: "Heading 1",
        basedOn: "Normal",
        next: "Normal",
        quickFormat: true,
        run: { size: 32, bold: true, font: "Calibri" },
        paragraph: { spacing: { before: 240, after: 120 }, outlineLevel: 0 },
      },
    ],
  },
  sections: [
    {
      properties: {
        page: {
          size: { width: 11906, height: 16838 },
          margin: { top: 1440, right: 1440, bottom: 1440, left: 1440 },
        },
      },
      headers: {
        default: new Header({
          children: [
            new Paragraph({
              children: [new TextRun({ text: "Smoke report", italics: true, size: 18 })],
            }),
          ],
        }),
      },
      footers: {
        default: new Footer({
          children: [
            new Paragraph({
              alignment: AlignmentType.CENTER,
              children: [new TextRun({ children: [PageNumber.CURRENT] })],
            }),
          ],
        }),
      },
      children: [
        new Paragraph({
          heading: HeadingLevel.HEADING_1,
          children: [new TextRun("Quarterly summary (smoke)")],
        }),
        new Paragraph({
          spacing: { after: 200 },
          children: [
            new TextRun(
              "Body text for make-docx smoke: one thesis per paragraph, A4 + DXA table.",
            ),
          ],
        }),
        new Table({
          width: { size: TABLE_W, type: WidthType.DXA },
          columnWidths: COLS,
          rows: [
            new TableRow({
              children: [
                cell("Item", COLS[0], { bold: true, fill: "D9E2F3" }),
                cell("Q2", COLS[1], { bold: true, fill: "D9E2F3" }),
                cell("Q3", COLS[2], { bold: true, fill: "D9E2F3" }),
              ],
            }),
            new TableRow({
              children: [
                cell("Revenue", COLS[0]),
                cell("120", COLS[1]),
                cell("145", COLS[2]),
              ],
            }),
            new TableRow({
              children: [
                cell("Margin %", COLS[0]),
                cell("18", COLS[1]),
                cell("21", COLS[2]),
              ],
            }),
          ],
        }),
      ],
    },
  ],
});

mkdirSync(outDir, { recursive: true });
const buf = await Packer.toBuffer(doc);
writeFileSync(outFile, buf);
console.log(`wrote ${outFile}`);

const checker = join(__dirname, "check_docx_tables.py");
const check = spawnSync("python", [checker, outFile], { encoding: "utf8" });
process.stdout.write(check.stdout || "");
process.stderr.write(check.stderr || "");
process.exit(check.status === 0 ? 0 : 1);

File: skills/make-xlsx/scripts/smoke_create.py
#!/usr/bin/env python3
"""Smoke: create a sample .xlsx and run check_xlsx_basics.py.

Usage (from repo / workspace root):
  python skills/make-xlsx/scripts/smoke_create.py
"""

from __future__ import annotations

import subprocess
import sys
from pathlib import Path

from openpyxl import Workbook
from openpyxl.styles import Alignment, Border, Font, PatternFill, Side

SCRIPT_DIR = Path(__file__).resolve().parent
WORKSPACE = SCRIPT_DIR.parents[2]
OUT = WORKSPACE / "outputs" / "documents" / "xlsx" / "smoke-product-mix.xlsx"


def main() -> int:
    wb = Workbook()
    ws = wb.active
    ws.title = "Summary"

    header_font = Font(name="Calibri", bold=True, color="1A1A1A")
    header_fill = PatternFill("solid", fgColor="D9E2F3")
    thin = Border(
        left=Side(style="thin", color="B0B0B0"),
        right=Side(style="thin", color="B0B0B0"),
        top=Side(style="thin", color="B0B0B0"),
        bottom=Side(style="thin", color="B0B0B0"),
    )

    headers = ["Item", "Amount ($)", "Share"]
    for col, text in enumerate(headers, start=1):
        cell = ws.cell(1, col, text)
        cell.font = header_font
        cell.fill = header_fill
        cell.alignment = Alignment(horizontal="center")
        cell.border = thin

    ws["A2"] = "Product A"
    ws["B2"] = 1200
    ws["A3"] = "Product B"
    ws["B3"] = 800
    ws["A4"] = "Total"
    ws["B4"] = "=SUM(B2:B3)"
    ws["C2"] = "=B2/$B$4"
    ws["C3"] = "=B3/$B$4"

    for coord in ("B2", "B3", "B4"):
        ws[coord].number_format = "$#,##0"
    for coord in ("C2", "C3"):
        ws[coord].number_format = "0.0%"

    ws.column_dimensions["A"].width = 18
    ws.column_dimensions["B"].width = 14
    ws.column_dimensions["C"].width = 10
    ws.freeze_panes = "A2"

    OUT.parent.mkdir(parents=True, exist_ok=True)
    wb.save(OUT)
    print(f"wrote {OUT}", flush=True)

    checker = SCRIPT_DIR / "check_xlsx_basics.py"
    proc = subprocess.run(
        [sys.executable, str(checker), str(OUT)],
        check=False,
    )
    return proc.returncode


if __name__ == "__main__":
    raise SystemExit(main())
